include/BareCpper/ATsamd/ExternalInterruptController_ATsamd5x.hpp

```diff
@@ -0,0 +1,182 @@
+#ifndef EXTERNAL_INTERRUPT_CONTROLLER_SAMD51_HPP
+#define EXTERNAL_INTERRUPT_CONTROLLER_SAMD51_HPP
+
+#if !__SAMD51__
+#  error "SAMD library error, please check and update accordingly."
+#endif
+
+#include <functional>
+
+#include "../Gpio.hpp"
+#include <sam.h>
+
+namespace BareCpper
+{
+  namespace SAMD51
+  {
+    /**
+     * Edge or level configuration for
+     * the external interrupt/event
+     * 
+     */
+    enum class ExternalInterruptSense : uint8_t
+    {
+      None  = 0x0,
+      Rise  = 0x1,
+      Fall  = 0x2,
+      Both  = 0x3,
+      High  = 0x4,
+      Low   = 0x5
+    };
+
+    enum class ExternalInterruptDebouncer : uint8_t
+    {
+      Off = 0xFF,
+      Prescaler2    = 0x0,
+      Prescaler4    = 0x1,
+      Prescaler8    = 0x2,
+      Prescaler16   = 0x3,
+      Prescaler32   = 0x4,
+      Prescaler64   = 0x5,
+      Prescaler128  = 0x6,
+      Prescaler256  = 0x7,
+    };
+
+    class ExternalInterruptController
+    {
+      public:
+        static constexpr uint8_t NumberChannels = 16;
+        /**
+         * Initialize the EIC
+         * 
+         */
+        void init()
+        {
+          // EIC MCLK is enabled by default
+          // Use GCLK1 for EIC (max is 100 MHz)
+          GCLK->PCHCTRL[EIC_GCLK_ID].reg = GCLK_PCHCTRL_GEN_GCLK1_Val | (1 << GCLK_PCHCTRL_CHEN_Pos);
```

Comment:
Design:  Possibly a future optimization, but the GCLK is not required for EIC except under filter, edge, and debounce interrupt modes. GCLK could be started based in `attachInterrupt()` (29.6.2.1 @ step 3)

Current 'Falling-edge' is used so clock is required by client code.

Reply:
Nice, we can change it when/as needed.